test/ecto/multi_test.exs

```diff
@@ -330,4 +271,72 @@ defmodule Ecto.MultiTest do
       Multi.new |> Multi.run(:run, fun) |> Multi.run(:run, fun)
     end
   end
+
+  test "merge with fun" do
+    changeset = Changeset.change(%Comment{})
+    multi =
+      Multi.new
+      |> Multi.insert(:insert, changeset)
+      |> Multi.merge(fn data ->
+        Multi.new |> Multi.update(:update, Changeset.change(data.insert))
+      end)
+
+    assert {:ok, data} = TestRepo.transaction(multi)
+    assert %Comment{} = data.insert
+    assert %Comment{} = data.update
+  end
+
+  test "merge with mfa" do
+    changeset = Changeset.change(%Comment{})
+    multi =
+      Multi.new
+      |> Multi.insert(:insert, changeset)
+      |> Multi.merge(__MODULE__, :multi, [])
+
+      assert {:ok, data} = TestRepo.transaction(multi)
+      assert %Comment{} = data.insert
+      assert %Comment{} = data.update
+  end
+
+  test "merge rollbacks on errors" do
+    error = fn _ -> {:error, :error} end
+    ok    = fn _ -> {:ok, :ok} end
+
+    multi =
+      Multi.new
+      |> Multi.run(:outside_ok, ok)
+      |> Multi.merge(fn _ ->
```

Comment:
We should also test that the merged operations run before any upcoming operation in the parent changeset. For example, we can add `outside_error` at the end and ensure it won't be triggered.